### DISCUSSION TITLE
Meson and cmake not finding system installed fltk using cmake

### DISCUSSION THREAD

**Discussion by andy5995** [fltk/fltk#1169] [General]
Partly discussed in https://github.com/fltk/fltk/issues/1148

@Albrecht-S:
>@KaruroChori First of all, if you use (or your meson build uses) CMake to build FLTK then you should NOT use fltk-config at all. fltk-config is a script that has been used (and can still be used) to build small programs as prototypes (fltk-config --compile) and since FLTK 1.4 you can also use multiple source files and link libraries.

>The other options of fltk-config can be used to request the compiler and flags (similar to what pkg-config does) but this is NOT intended (nor is it recommended) to be used within CMake and it is not a replacement for pkg-config.

Building the subproject works fine, however, when fltk is installed on Linux (via pacman), or the MacOS runner (via brew), @mesonbuild is not finding fltk via pkg-config or cmake.

With meson, usually first the dependency() method is used to find the compile and link flags. If that doesn't work, find_library() can be used. On Manjaro, dependency() didn't work, but find_library() worked. In the MacOS runner, neither one worked.

From the command line, cmake can find package info and output info in a manner similar to pkg-config. For example:

```
$ cmake --find-package -DNAME=ZLIB -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=COMPILE
-I/usr/include 
$ cmake --find-package -DNAME=ZLIB -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=LINK
/usr/lib/libz.so 
$ cmake --find-package -DNAME=CURL -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=LINK
/usr/lib/libcurl.so 
$ cmake --find-package -DNAME=CURL -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=COMPILE
-I/usr/include
```

But when trying to find fltk:

```
$ cmake --find-package -DNAME=FLTK -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=COMPILE
CMake Error at /usr/share/cmake/Modules/CheckLibraryExists.cmake:75 (message):
  CHECK_FUNCTION_EXISTS needs either C or CXX language enabled
Call Stack (most recent call first):
  /usr/share/cmake/Modules/FindX11.cmake:693 (check_library_exists)
  /usr/share/cmake/Modules/FindFLTK.cmake:102 (include)
  /usr/share/cmake/Modules/CMakeFindPackageMode.cmake:182 (find_package)


FLTK not found.
CMake Error: Run 'cmake --help' for all supported options.
```

**Comment by andy5995** [comment]
I've found a little more information and still investigating.

I see you do ship 'FLTKConfig.cmake' and it's installed when 'make install' is run (or from a distro package).

cmake should be able to use this file using [config search mode](https://cmake.org/cmake/help/latest/command/find_package.html#id7)

```
In this mode, CMake searches for a file called <lowercasePackageName>-config.cmake or
<PackageName>Config.cmake. It will also look for <lowercasePackageName>-config-version.cmake
or <PackageName>ConfigVersion.cmake if version details were specified
(see [Config Mode Version Selection](https://cmake.org/cmake/help/latest/command/find_package.html#version-selection) for an explanation of how these separate version files are used).
...
```

There are some additional ways to look for a dependency using cmake with meson, noted [here in the docs](https://mesonbuild.com/Dependencies.html#cmake)

**Reply by Albrecht-S** [reply]
See my comment below: meson doesn't seem to use FLTKConfig.cmake which is a pity.

**Comment by Albrecht-S** [comment]
@andy5995 Thanks for the report, but I'm afraid this is not a problem the FLTK team can resolve. I'm open for all suggestions for improvements though. For details see below.

You wrote:
> I see you do ship 'FLTKConfig.cmake' and it's installed when 'make install' is run (or from a distro package).

Here's an important point: `FLTKConfig.cmake` is only generated and installed by `make install` if the build was done using CMake. You also wrote "installed ... from a distro package": this would require that the distro built FLTK using CMake and installed it properly.

Your example to find FLTK with CMake's find_package() from the commandline works on my system well:
```
$ cmake --find-package -DNAME=FLTK -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=COMPILE
-I/home/albrecht/fltk/1.4/debug/include -I/usr/include 
```
well, so far, so good, however I admit that this doesn't seem right:
```
$ cmake --find-package -DNAME=FLTK -DCOMPILER_ID=GNU -DLANGUAGE=C -DMODE=LINK
-Wl,-rpath,/home/albrecht/fltk/1.4/debug/lib  /home/albrecht/fltk/1.4/debug/lib/libfltk_images.so  /home/albrecht/fltk/1.4/debug/lib/libfltk_forms.so  /home/albrecht/fltk/1.4/debug/lib/libfltk_gl.so  -lOPENGL_gl_LIBRARY-NOTFOUND  /home/albrecht/fltk/1.4/debug/lib/libfltk.so  -lFLTK_MATH_LIBRARY-NOTFOUND
```

We should also note that FLTK uses neither `OPENGL_gl_LIBRARY` nor `FLTK_MATH_LIBRARY` in its CMake code or anywhere else. I have no idea how CMake achieves this output.

However, the CMake docs state that this mode of operation is not reliable and should not be used:
> It searches a package using [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) and prints the resulting flags to stdout. This can be used instead of pkg-config to find installed libraries **in plain Makefile-based projects or in autoconf-based projects** (via share/aclocal/cmake.m4).
> **_Note:_** This mode is not well-supported **due to some technical limitations**. It is kept for compatibility but should not be used in new projects. 

This all lets me conclude that your system installation "_on Linux (via pacman), or the MacOS runner (via brew)_" doesn't work as it should.

That said: current FLTK requires **CMake CONFIG** mode to find the libraries as expected. This is documented in README.CMake.txt. The old-style `FindFLTK.cmake` module bundled with CMake does NOT work as expected. It **may** work though under certain circumstances by using some of the provided CMake variables but this is not guaranteed.

Note that the CMake people discourage using FindXXX modules but recommend to use `find_package()` in `CONFIG` mode always.

For a detailed analysis what's going on on your system or build we don't have enough data about your system installation.

I can try to test what `brew install fltk` does and whether they provide `FLTKConfig.cmake` at least in their installation of FLTK 1.4.0 but this can take a while. Maybe you can do this easier...

Sorry for the long reply. Maybe you can provide more information but I'm not sure if this helps because I don't use `meson` and don't know how to figure out what's going wrong. Looking forward to seeing more info and willing to help if possible...

**Reply by Albrecht-S** [reply]
PS: one way to build FLTK as a dependency in a `CMake` project is maybe using `FetchContent` in your (super)build and use the generated `FLTKConfig.cmake` file of that (sub-)build. There's an example in chapter 3.6: _"Building a Program Using CMake's FetchContent Module"_ in `README.CMake.txt`. HTH.

**Reply by andy5995** [reply]
> You wrote:
> 
> > I see you do ship 'FLTKConfig.cmake' and it's installed when 'make install' is run (or from a distro package).
> 
> Here's an important point: `FLTKConfig.cmake` is only generated and installed by `make install` if the build was done using CMake. You also wrote "installed ... from a distro package": this would require that the distro built FLTK using CMake and installed it properly.

Yes, it's standard that distro packagers build projects with the projects primary build system. And I've verified that FLTKConfig.cmake has been installed on Debian trixie and Manjaro. On Manjaro the path is '/usr/share/fltk/' and on Trixie it's `/usr/lib/fltk/`.

The Debian packaging rules are at https://salsa.debian.org/ucko/fltk1.4/-/blob/main/debian/rules?ref_type=heads and I don't see they make any changes to the location of FLTKConfig.cmake 

Same for the [Manjaro/Arch PKGBUILD](https://gitlab.archlinux.org/archlinux/packaging/packages/fltk/-/blob/main/PKGBUILD?ref_type=heads)

And same for the [macos/brew formula](https://github.com/Homebrew/homebrew-core/blob/34b2b819b4f857ea2c5d81e0b6472ad90c22bd7c/Formula/f/fltk.rb) 

I do have a correction to make. This is the output in the MacOS runner:

```
Found CMake: /opt/homebrew/bin/cmake (3.31.1)
Dependency fltk found: NO. Unknown version, but need: ['>=1.4']
Run-time dependency fltk found: NO (tried pkgconfig, framework and cmake)
```

> 
> Your example to find FLTK with CMake's find_package() from the commandline works on my system well:
> 

That's interesting. Do you know how cmake is finding the information? Which distro are you using. Do you have any CMAKE environmental variables set?
> However, the CMake docs state that this mode of operation is not reliable and should not be used:
> 
>
> > **_Note:_** This mode is not well-supported **due to some technical limitations**. It is kept for compatibility but should not be used in new projects.

It sounds like the cmake config files should not be the sole way that the dependency information is found, and that a reasonable option might be for the fltk cmake  to create a .pc file during installation.

[How to generate .pc (pkg-config) file supporting –prefix of the cmake –install?](https://discourse.cmake.org/t/how-to-generate-pc-pkg-config-file-supporting-prefix-of-the-cmake-install/4109)

> 
> Note that the CMake people discourage using FindXXX modules but recommend to use `find_package()` in `CONFIG` mode always.

> 
> For a detailed analysis what's going on on your system or build we don't have enough data about your system installation.

> Sorry for the long reply. Maybe you can provide more information but I'm not sure if this helps because I don't use `meson` and don't know how to figure out what's going wrong. Looking forward to seeing more info and willing to help if possible...

No need to apologize. :) I appreciate your thoroughness.

I did some more checking and found there is [a meson ticket for this](https://github.com/mesonbuild/meson/issues/13228) as well. I'll update that ticket and maybe there's some extra kwargs that can be added so meson can find fltk.

But here's the testing I did, a small project:

```
#include <stdio.h>

int main() {
    printf("Hello, World!\n");
    return 0;
}
```
## cmake

```
cmake_minimum_required(VERSION 3.15)
project(SimpleProject)

# Specify C standard
set(CMAKE_C_STANDARD 99)

# Find and link FLTK
find_package(FLTK REQUIRED)
include_directories(${FLTK_INCLUDE_DIRS})

# Find and link fmt
find_package(fmt REQUIRED)

# Add the executable
add_executable(simple_project main.c)

# Link libraries
target_link_libraries(simple_project PRIVATE ${FLTK_LIBRARIES} fmt::fmt)
```

fltk was found:

```
-- Looking for IceConnectionNumber in ICE - found
-- Found FLTK: /usr/lib/libfltk_images.so;/usr/lib/libfltk_forms.so;/usr/lib/libfltk_gl.so;/usr/lib/libfltk.so
-- Configuring done (0.9s)
-- Generating done (0.0s)
-- Build files have been written to: /home/andy/src/cmake_simple/build
```

## Meson

```
project('simple_project', 'c',
  version: '1.0',
  meson_version: '>= 0.55'
)

# Define the source files
sources = ['main.c']

# Find FLTK dependency
fltk_dep = dependency('fltk', required: false)

# Find fmt dependency
fmt_dep = dependency('fmt', required: true)

# Build the executable
executable('simple_project',
  sources,
  dependencies: [fltk_dep, fmt_dep]
)
```

Output:

```
Found pkg-config: YES (/usr/bin/pkg-config) 2.3.0
Found CMake: /usr/bin/cmake (3.31.1)
Run-time dependency fltk found: NO (tried pkgconfig and cmake)
Run-time dependency fmt found: YES 11.0.2
Build targets in project: 1
```

I think you asked about my use-case... I'm a contributor on vs-fltk and [saw this issue came up](https://github.com/KaruroChori/vs-fltk/pull/67) and wanted to discuss or solve it. :)

**Reply by KaruroChori** [reply]
> I do have a correction to make. This is the output in the MacOS runner:
>
>      Found CMake: /opt/homebrew/bin/cmake (3.31.1)
>      Dependency fltk found: NO. Unknown version, but need: ['>=1.4']  
>      Run-time dependency fltk found: NO (tried pkgconfig, framework and cmake)

Yes, somehow by removing the version requirement it is detected, but I was unable to find images and forms. Why the version is not exposed to be matched is a mistery.

**Reply by Albrecht-S** [reply]
> Why the version is not exposed to be matched is a mistery.

Maybe the same cause I found and commented above.

Can you try searching for `FLTKConfig.cmake` in the homebrew install tree and then
```
$ touch CMakeCache.txt
```
in the same folder where you found it - **if it doesn't exist yet**?

Then try again...

Note that you should have at least FLTK 1.4.0 but I read that homebrew updated to 1.4.1 already, I'm not on my mac now, hence I can't verify this.

**Reply by KaruroChori** [reply]
Sadly I don't have a mac myself, I only use that platform via github actions for automatic testing/building. Performing arbitrary operations via a real time shell is not really feasible.

**Reply by Albrecht-S** [reply]
Ah, OK, then this request should be addressed to @andy5995 ...

**Comment by andy5995** [comment]
As I note [here](https://github.com/mesonbuild/meson/issues/13228#issuecomment-2543107323), meson can find fltk using it's find_library() method.

**Reply by Albrecht-S** [reply]
Short reply first: in the CMake world using `find_library()` is inferior to using the "Modern CMake" approach that uses the namespaced target names `fltk::fltk`, `fltk::images` etc.. This requires searching in `CONFIG` mode as described in `README.CMake.txt` and provides a CMake build with all dependencies.

I assume (!) that a bare `find_library()` in meson would only find the lib(s) but not any required include directories nor other dependencies. But I don't know meson...

(more to follow in reply to other comments)

**Comment by Albrecht-S** [comment]
OK, this one is going to be long again and I'll split it in multiple replies / comments.

> I've verified that FLTKConfig.cmake has been installed on Debian trixie and Manjaro.

According to [Repology](https://repology.org/project/fltk/versions) even Manjaro Unstable uses FLTK 1.3.9, and Debian Trixie 1.3.8. Both versions lack the latest addition `FLTKConfigVersion.cmake` (since 1.3.10 and 1.4.0, release on Nov. 16/18 resp.). This can be an issue if more than one FLTK version is installed or if your build system requests a particular version. Without `FLTKConfigVersion.cmake` CMake doesn't accept version 1.3.8/1.3.9 because no version is defined. A workaround would be not to request any version if 1.3.x (x < 10) is installed and sufficient.

The Debian maintainer agreed to provide `1.3.10` and `1.4.0` or higher (current version: 1.4.1) in the future Trixie release (but I can't tell yet when this will appear in `Debian Testing` alias Trixie (as you certainly know Trixie has not yet been released, it's currently Debian Testing).

**Reply by Albrecht-S** [reply]
Pausing. Dinner break. I'm not sure if I can continue tonight, maybe tomorrow...

**Reply by KaruroChori** [reply]
No worries, I still have to read the rest of the conversation :). Have a good dinner!
And thanks for your time so far

**Comment by Albrecht-S** [comment]
### Summary
1. Bad news: `meson` 1.0.1 does not (yet?) use `CMake` in `CONFIG` mode.
2. Good news: I made progress and found a working `meson` config. Working for me at least.

#### Test report

Well, it took some time to learn the basics of meson. My system is Debian 12 (Bookworm) and provides cmake 3.25.1 and meson 1.0.1, hence I didn't test with newer meson versions. 

It was hard to find the right meson docs, for instance "what is a kwarg"? They're writing much about `kwargs` but don't seem to explain what this means. Finally I found that it means "keyword arg" (I hope that's correct). But that's not important for us now...

1. I looked at the internal "CMakeLists.txt" created by meson which is obviously used to find dependencies with CMake. It's in `<build-dir>/meson-private/cmake_FLTK/CMakeLists.txt`. This file contains:
```
  if ("${VERSION}" STREQUAL "")
    find_package("${NAME}" QUIET COMPONENTS ${COMPS})
  else()
    find_package("${NAME}" "${VERSION}" QUIET COMPONENTS ${COMPS})
  endif()
```

which is not what we recommend in our `README.CMake.txt` for FLTK 1.3 and 1.4. This uses the outdated `FindFLTK.cmake` "module" bundled with CMake - unless you create your own because the keyword "CONFIG" is missing. _[Side note: FLTK doesn't provide `FindFLTK.cmake` because this is no longer "state of the art" anyway because it provides libraries and flags in `CMake` variables rather than CMake 'targets' as "modern CMake" demands.]_ I could write much more about this, but this would be OT.

Then I tried to edit this file in the hope that I could make it work temporarily - but meson overwrites it again before using it, hence this approach failed.

2. Anyway, I found a working config:
```meson
# This meson project works with meson 1.0.1 and FLTK 1.3.8 and 1.4.1
#
# Problem: meson needs some help to find FLTK because it does
# not (yet?) use CMake's CONFIG mode but relies on FindFLTK.cmake.
# This old-style CMake file is known not to work with "modern CMake".
#
# Disclaimer: I know only basics about 'meson'. The meson config below
# is very likely not "good meson style" (created by trial and error).

# Note: in the tested configuration my PATH includes the FLTK
#       installation folder where 'fluid' is located:
#   (a) /path/to/fltk/1.3.8/debug/bin   # FLTK 1.3.8
#   (a) /path/to/fltk/1.4.1/debug/bin   # FLTK 1.4.1
#
# I'm not using a system-installed FLTK version!

project('simple_project', 'cpp',
  version: '1.0',
  meson_version: '>= 0.56'
)

# Define the source files
sources = ['hello.cxx']

# Find FLTK dependency
fltk_dep = dependency('FLTK', method : 'cmake', required: true)

# Find more FLTK dependencies (temporary solution)
#  - why do we need this at all ? Note: tested with *old* FLTK version 1.3.8
#  - if we need it, couldn't this be simplified ? More dependencies per line ??

### These dependencies are required for FLTK 1.3.8
### ... and can be commented out for FLTK 1.4

xrender_dep  = dependency('xrender')
xcursor_dep  = dependency('xcursor')
xinerama_dep = dependency('xinerama')
xfixes_dep   = dependency('xfixes')
xft_dep      = dependency('xft')
fc_dep       = dependency('fontconfig')

### This dependency is required for FLTK 1.4.1
### ... and can be commented out for FLTK 1.3.8
### Note: works with FLTK 1.4.1 using X11 and Wayland!

cairo_dep    = dependency('cairo') # needed for Wayland

# Build the executable

executable('hello',
  sources,
  dependencies: [fltk_dep,                  # FLTK 1.3.8 or 1.4.1
  xcursor_dep, xrender_dep, xinerama_dep,   # FLTK 1.3.8 only
  xfixes_dep, xft_dep, fc_dep,              # FLTK 1.3.8 only
  cairo_dep]                                # FLTK 1.4.1 only
)
```
It's long and full of comments, please read yourself. This is not yet a proper solution but it reflects why `FindFLTK.cmake` is no longer recommended. Everybody should use CMake's `CONFIG` mode to find libraries and use the **_CMake targets_** rather than CMake variables. I'm aware that adding dependencies manually is not a real solution, but it's a workaround for now.

#### Details

The only way I found **_on my development system_** - where multiple FLTK versions are installed, but none in the system directories - was to use the PATH to direct meson + cmake to the correct installation folder. This allowed me to switch from 1.3.8 (intentionally an older version which is installed on many Linux systems) and 1.4.1, the brand new FLTK release. Note that '`/debug/`' is part of **_my installation path_** per FLTK version (I have also '/release'/ and more for testing purposes). But when the path to the respective `fluid` file was set, then meson found the corresponding FLTK version. BTW, this is the same with "classic" CMake projects, i.e. if you don't use "modern CMake".

The next problem was that this didn't find all required libraries. I didn't investigate further, it is very likely the fault of `FindFLTK.cmake` but it is also possible that this is an FLTK issue. I'll check this further later. Maybe.

Now, after being that far, I tried to link the missing libraries "manually" (see my comments in the meson file). I'd appreciate if someone could tell me how to simplify this: I had to define one dependency for each library and finally add each dependency to the `dependencies`. This looks like much redundant work (indeed I forgot to add or remove one of the things while testing). Can this be simplified? Please reply in simple wording as in "meson for dummies".

### Conclusion:
So far my experiments and my "working" meson build config. For now it seems possible to find FLTK by providing the correct PATH (to `fluid`) or something similar which needs - AFAICT - to add libraries manually, unfortunately.

My suggestion would be to improve meson that it can handle `CMake` searches in `CONFIG` mode which would also find the correct FLTK version (since FLTK 1.3.10 and/or 1.4.0, using `FLTKConfigVersion.cmake`). This is the "modern CMake" approach and definitely the future.

My next comment will be about `pkgconf`...

**Reply by KaruroChori** [reply]
I am also going to split my replies because there is quite a bit to digest :).

> Well, it took some time to learn the basics of meson. My system is Debian 12 (Bookworm) and provides cmake 3.25.1 and meson 1.0.1, hence I didn't test with newer meson versions.

I would strongly recommend to upgrade it, because version 1.0.1 is almost 2 years old.  
It might be very hard to judge if some of the rough edges being observed are because of bugs already solved, or some which are not :D.
It is distributed as a python package, so it is as easy as `pip install --user meson`.
Make sure if you installed a version with `apt` before to uninstall it.

**Reply by Albrecht-S** [reply]
Thanks. For now I'd prefer to stick with 1.0.1 because I assume that other users may also have that (old) version. But I'll keep your advice in mind...

**Reply by andy5995** [reply]
@Albrecht-S I regret I'm not able to read through all of this. I appreciate your following up and even giving it a try with meson.

> Thanks. For now I'd prefer to stick with 1.0.1 because I assume that other users may also have that (old) version. But I'll keep your advice in mind...

Yes, it's hard to know if a majority of users are using a newer version. On the other hand, any problems you find while testing with the older version may not be relevant enough to be reported to the meson developers (due to the age of meson 1.0.1).

While testing some meson patches, I've used `meson.py` from the source root of the meson repo (therefore, I didn't install using pip).

**Reply by Albrecht-S** [reply]
Thanks, @andy5995, I'm aware of this but I wanted to find out how it worked with the "current" version on Debian Stable (Bookworm). Debian Trixie will likely be released in June 2025. As it looks now I found a working solution, and it's great that this works even with the old meson version. BTW, Debian Testing still has meson 1.0.1 but I assume it will be upgraded before the release. You as meson users could maybe file an issue...

**Reply by andy5995** [reply]
> Thanks, @andy5995, I'm aware of this but I wanted to find out how it worked with the "current" version on Debian Stable (Bookworm). Debian Trixie will likely be released in June 2025. As it looks now I found a working solution, and it's great that this works even with the old meson version. BTW, Debian Testing still has meson 1.0.1 but I assume it will be upgraded before the release. You as meson users could maybe file an issue...

Debian [testing/trixie has 1.6.0](https://packages.debian.org/trixie/meson)

By the way, if it's any more convenient for you, the vs-fltk project has a trixie build environment with meson already installed: https://github.com/KaruroChori/vs-fltk/pkgs/container/vs-fltk/320734774?tag=build-env

**Reply by Albrecht-S** [reply]
> Debian [testing/trixie has 1.6.0](https://packages.debian.org/trixie/meson)

OK, I missed that. I was working in a VM which is actually also Debian Stable (but I installed the system version of FLTK for my tests). I have another VM with Testing/Trixie but I need to check it there in action. Sorry for the confusion.

Thanks for the build env hint, but I have my own Trixie VM as written above. And, BTW, I can't test all FLTK using projects like vs-fltk although I would like to do more like this. I'm driven by making FLTK better and usable in all environments, and `meson` is one I'm personally interested in but I had only done minimal tests before and forgot almost everything. ;-)

**Comment by Albrecht-S** [comment]
### More good news
I made even more progress and found a much better working solution. Note that my test project worked on Debian Testing (Trixie) :+1: OOTB with the system-installed FLTK version (1.3.8).

I'm still investigating the root cause. It seems that `find_package(FLTK)` wants to find a `CMakeCache.txt` file in the same folder where `FLTKConfig.cmake` is found. Even an empty file suffices, and then it works well using `CONFIG` mode. Wow, that's surprising and impressing! I'll let you know when I found more.

**_Correction:_** all tests that mention Debian 13 (Testing/Trixie) have been done on Debian 12 (Bookworm).

**Comment by Albrecht-S** [comment]
OK, I think I got it, here's a more or less final solution:
1. FLTK should install either an empty or a more useful `CMakeCache.txt` file in the folder where `FLTKConfig.cmake` is installed. As long this is not the case users could `touch CMakeCache.txt` to create an empty file. This needs more investigation, I was not aware that this `CMakeCache.txt` is of any use. I noticed that Debian has it installed.
2. With this in place setting the environment variable `FLTK_DIR` to point at the location of the `FLTKConfig.cmake` file enables `meson` to find FLTK in that location. Even requesting a version works (since FLTK 1.3.10/1.4.0).
~3. I assume ... (forgot what I wanted to say)~

#### Example meson.build file
```meson
project('hello', 'cpp',
  version: '1.0',
  meson_version: '>= 0.56'
)

# Define the source files
sources = ['hello.cxx']

# Find FLTK dependency
fltk_dep = dependency('FLTK',
                      method: 'cmake',
                      version: '>=1.4.1',
                      required: true)

# Build the executable
executable('hello', sources, dependencies: fltk_dep)
```
#### Example source file `hello.cxx`
```C++
#include <FL/Fl.H>
#include <FL/Fl_Window.H>
#include <FL/Fl_Box.H>

int main(int argc, char **argv) {
  char version[120];
  sprintf(version, "Hello,\nFLTK %d.%d.%d", FL_MAJOR_VERSION, FL_MINOR_VERSION, FL_PATCH_VERSION);
  Fl_Window *window = new Fl_Window(340, 180);
  Fl_Box *box = new Fl_Box(20, 40, 300, 100, version);
  box->box(FL_UP_BOX);
  box->labelfont(FL_BOLD + FL_ITALIC);
  box->labelsize(30);
  box->labeltype(FL_SHADOW_LABEL);
  window->end();
  window->show(argc, argv);
  return Fl::run();
}
```

#### meson setup:
```
$ FLTK_DIR=/home/albrecht/fltk/1.4/debug/share/fltk meson setup build
The Meson build system
Version: 1.0.1
Source dir: /git/fltk/projects/meson
Build dir: /git/fltk/projects/meson/build
Build type: native build
Project name: simple_project
Project version: 1.0
C++ compiler for the host machine: c++ (gcc 12.2.0 "c++ (Debian 12.2.0-14) 12.2.0")
C++ linker for the host machine: c++ ld.bfd 2.40
Host machine cpu family: x86_64
Host machine cpu: x86_64
Found CMake: /usr/bin/cmake (3.25.1)
WARNING: CMake: Dependency cairo for FLTK was not found
Run-time dependency fltk found: YES 1.4.1
Build targets in project: 1

Found ninja-1.11.1 at /usr/bin/ninja
```

#### Result on Debian Bookworm with self-built FLTK 1.4.1

![image](https://github.com/user-attachments/assets/b15b7c47-5b35-483f-b0d4-8677a61b3362)

#### Result on Debian Testing/Trixie with system-installed FLTK 1.3.8

![image](https://github.com/user-attachments/assets/0d4642c1-717a-4131-b3a0-a8dc5f5abf80)

Note that this worked on Debian Trixie w/o specifying FLTK_DIR, as I would have expected (because FLTKConfig.cmake is installed in a system folder). I didn't specify a min version though.

**_Correction:_** all tests above that mention Debian 13 (Testing/Trixie) have been done on Debian 12 (Bookworm).

Please let me know if this works for you. TIA.

**Reply by KaruroChori** [reply]
I tested it with 1.3.8. The crucial point is the name of the library being `FLTK` and not `fltk`. Most names are lower case so I always defaulted to that, but it was the major reason for it not working. Plus few more factors. The `cmake` method can also be dropped, as it will automatically use that if pkgconfig fails.

**Reply by Albrecht-S** [reply]
> The crucial point is the name of the library being FLTK and not fltk

Yes, that has historical reasons, I would like to rename it to 'fltk' but this would have some side effects. I wonder though why meson (or CMake?) is case sensitive here. I might have tried both but it worked for me with FLTK.

> The cmake method can also be dropped, ...

OK, good to know that it can be simplified. The reason I kept it in my tests was that I had used `cmake_args` in previous versions but it turned out that I didn't need them (e.g. '-D FLTK_DIR=/path-to-FLTKConfig.cmake' which works as well if needed but I prefer the commandline).

**Reply by KaruroChori** [reply]
~~The only question left before I test with 1.4.1 is how do I also take the deps for images/forms/gl? Because all naming schemes I tested failed so far.~~

Found it, it was resolved in https://github.com/mesonbuild/meson/issues/13228, even if it might not be the best option. 
```
dependency('FLTK', modules: [....])
```

So in theory the only real problem left (except for brew which has to be tested on its own) it the lack of that `CMakeCache.txt` which is added by the debian maintainers for good reasons. But for testing purposes it can just be added after installing.

**Reply by Albrecht-S** [reply]
Yes, this works with FLTK 1.4 and higher as well. To clarify: I tested
```
fltk_dep = dependency('FLTK', version: '>=1.4', required: true,
                      modules: ['fltk::fltk', 'fltk::images', 'fltk::gl', 'fltk::forms'] )
```
successfully, although there were some warnings. This kind of name-spaced module names is new since FLTK 1.4.0 ("modern CMake").

The warnings seem to be benign and the build works well.

**Reply by KaruroChori** [reply]
For macos it looks like the only way to make it work with the brew package is something like `libfltk_dep = dependency('fltk', required: true)` explicitly lowercase and no version number specified or the match always fail.  
I don't know if this is a bug of meson or of if brew is packaging it wrong to be honest. I just wanted to report it so that it is documented.  
My understanding is that the package resolution unlike on linux distros is not via `cmake` but via `framework` which must be something apple specific.

**Reply by Albrecht-S** [reply]
Oh, that's surprising - thanks for the report, it's appreciated.

So if this is all true: Linux needs 'FLTK', macOS/homebrew needs 'fltk', then this would really be bad. From my POV a `CMakeLists.txt` should ideally be cross-platform, and so should a `meson.build` file.

Are maybe all dependencies (libs) in homebrew lowercase?

I don't have meson on my Mac but I'll try to install it and test with the homebrew version of FLTK and let you know what I find out.

FLTK does not (yet) provide a macOS `framework`, this is planned for the future. Such a framework is a special kind of "bundle" (a directory with contents) in macOS which contains multiple files and can be installed and moved as a whole. The FLTK framework would likely contain all the separate libs, header files, maybe its own dependencies, etc..

**Comment by Albrecht-S** [comment]
OK, I think we're done with this discussion thread, and I hope it helped.

I'm going to investigate this `CMakeCache.txt` issue (why is it needed, what should it contain?) and I'll add this for FLTK 1.4.2 if I find a satisfying solution - hopefully I will. Right now I feel uncomfortable to add this file as long as I don't know what it is really good for. I'll dig into it.

Maybe it's worth adding something to our docs about "Using FLTK with Meson".

**Comment by Albrecht-S** [comment]
A note about `pkgconf` for FLTK: I thought about it some time ago but I came to the conclusion that it's not the right time to add it before FLTK 1.4.0 was released. I'm still not sure if it is useful at all and its complexity can't be underestimated although it looks trivial at a first glance. FLTK's flexibility to build several optional libs makes it challenging and to get their dependencies right (if optional libs are built or not) doesn't sound trivial - and I'm the main CMake developer for FLTK, I know what I'm talking about. ;-)

As previously announced we won't update the `autoconf`/`Makefile` build system anymore in FLTK 1.4.x and it will be removed in 1.5.0. Therefore `pkgconf` support will be delayed until we work on FLTK 1.5.0 and there's no guarantee that it will ever be done.

For now a working replacement for `pkgconf` is obviously `fltk-config` and I believe that `meson` could also apply it if there were the rules to use it, but that's only from scanning the docs quickly...

**Reply by KaruroChori** [reply]
> For now a working replacement for pkgconf is obviously fltk-config and I believe that meson could also apply it if there were the rules to use it, but that's only from scanning the docs quickly...

Yes,  but that would require to build a custom plugin for meson just for one dependency, or a lot of code to handle it in the meson.build of each project.  
Realistically, there is a functioning cmake interface; this is enough as we understood what was needed to make it work as expected.
